src/windows.rs

```diff
@@ -86,238 +121,318 @@ impl crate::Brightness for Brightness {
+        })
     }
 
     async fn set(&mut self, percentage: u32) -> Result<(), Error> {
-        let ioctl_query = ioctl_query_supported_brightness(self);
-        match ioctl_query {
-            Ok(levels) => {
-                let new_value = levels.get_nearest(percentage);
-                Ok(ioctl_set_display_brightness(self, new_value)?)
-            }
-            Err(e) if is_not_found(&e) => {
-                let new_value =
-                    ddcci_get_monitor_brightness(self)?.percentage_to_current(percentage);
-                Ok(ddcci_set_monitor_brightness(self, new_value)?)
-            }
-            Err(e) => Err(SysError::IoctlQuerySupportedBrightnessFailed {
-                device_name: self.device_name.clone(),
-                source: e,
-            }
-            .into()),
-        }
+        Ok(if self.is_internal() {
+            let supported = ioctl_query_supported_brightness(self)?;
+            let new_value = supported.get_nearest(percentage);
+            ioctl_set_display_brightness(self, new_value)?;
+        } else {
+            let current = ddcci_get_monitor_brightness(self)?;
+            let new_value = current.percentage_to_current(percentage);
+            ddcci_set_monitor_brightness(self, new_value)?;
+        })
     }
 }
 
 pub fn brightness_devices() -> impl Stream<Item = Result<Brightness, SysError>> {
+    unsafe {
+        let device_info_map = match get_device_info_map() {
+            Ok(info) => info,
+            Err(e) => return futures::stream::once(ready(Err(e))).left_stream(),
+        };
+        let hmonitors = match enum_display_monitors() {
+            Ok(monitors) => monitors,
+            Err(e) => return futures::stream::once(ready(Err(e))).left_stream(),
+        };
+        futures::stream::iter(hmonitors.into_iter().flat_map(move |hmonitor| {
+            let physical_monitors = match get_physical_monitors_from_hmonitor(hmonitor) {
+                Ok(p) => p,
+                Err(e) => return vec![Err(e)],
+            };
+            let display_devices = match get_display_devices_from_hmonitor(hmonitor) {
+                Ok(p) => p,
+                Err(e) => return vec![Err(e)],
+            };
+            if display_devices.len() != physical_monitors.len() {
+                // There doesn't seem to be any way to directly associate a physical monitor
+                // handle with the equivalent display device, other than by array indexing
+                // https://stackoverflow.com/questions/63095216/how-to-associate-physical-monitor-with-monitor-deviceid
+                return vec![Err(SysError::EnumerationMismatch)];
+            }
+            physical_monitors
+                .into_iter()
+                .zip(display_devices)
+                .filter_map(|(physical_monitor, mut display_device)| {
+                    let handle = match get_file_handle_for_display_device(&mut display_device) {
```

Comment:
It looks like it's possible for this handle to be leaked (there are exit points before it is wrapped in `WrappedFileHandle`). I suggest making `get_file_handle_for_display_device` return `Option<Result<WrappedFileHandle, SysError>>` to avoid this issue.

Reply:
Yep that is a good catch, have fixed as you suggested.

Reply:
Thank you!